pkg/controller/garbagecollector/codec.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type compressingCodec struct {
+	dynamic.Codec
+}
+
+func NewCompressingCodec() compressingCodec {
+	return compressingCodec{dynamic.Codec{}}
+}
+
+// Decode calls the embedded Codec to decode data. If the decoded object is
+// runtime.Unstructured or runtime.UnstructuredList, Decode only keep the
+// APIVersion, Kind, and ObjectMeta fields.
+func (c compressingCodec) Decode(data []byte, gvk *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
+	ret, gvk, err := c.Codec.Decode(data, gvk, into)
```

Comment:
@smarterclayton I took a closer look at the cache code. If we want to transform the object in the cache layer, I'll need to wrap the delta_fifo, transform the object in the Add, Update, Delete, Get, Replace methods. The code duplication makes the code dirtier than I thought. Is this what you have in mind?

For the codec solution, if you are worried about I hard coded the metadata fields name, I can always change to use the meta accessor, but it's kind of pointless, because we are hard coding [fields names](https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/types.go#L333) in the accessor anyway...